crates/sui-storage/src/event_store/sql.rs

```diff
@@ -370,74 +346,69 @@ const MAX_INSERT_BATCH: usize = 1000;
 #[async_trait]
 impl EventStore for SqlEventStore {
     #[instrument(level = "debug", skip_all, err)]
-    async fn add_events(&self, events: &[EventEnvelope]) -> Result<u64, SuiError> {
-        let mut cur_seq = self.seq_num.load(Ordering::Acquire);
-        let initial_seq = cur_seq;
+    async fn add_tx_events(&self, events: &[EventEnvelope]) -> Result<u64, SuiError> {
         let mut rows_affected = 0;
 
         if events.is_empty() {
             return Ok(0);
         }
 
-        let mut start_index = 0;
-        let mut end_index = 0;
-
-        // Insert event in batches, skipping over events that have lower or decreasing sequence numbers
-        while end_index < events.len() {
-            // Skip events that have a lower sequence number. They've been seen before.
-            if events[start_index].seq_num < cur_seq {
-                start_index += 1;
-                end_index += 1;
-                continue;
-            }
-
-            let final_index = (start_index + MAX_INSERT_BATCH).min(events.len());
-            // Keep going while the sequence number is not decreasing
-            while end_index < final_index && events[end_index].seq_num >= cur_seq {
-                cur_seq = events[end_index].seq_num;
-                end_index += 1;
-            }
-
-            let mut query_builder = QueryBuilder::new(SQL_INSERT_TX);
-            query_builder.push_values(&events[start_index..end_index], |mut b, event| {
-                let event_type = EventType::from(&event.event);
-                let sender = event.event.sender().map(|sender| sender.to_vec());
-                let move_event_name = event.event.move_event_name();
-                b.push_bind(event.timestamp as i64)
-                    .push_bind(event.seq_num as i64)
-                    .push_bind(event.tx_digest.map(|txd| txd.to_bytes()))
-                    .push_bind(event_type as u16)
-                    .push_bind(event.event.package_id().map(|pid| pid.to_vec()))
-                    .push_bind(event.event.module_name())
-                    .push_bind(event.event.object_id().map(|id| id.to_vec()))
-                    .push_bind(Self::event_to_json(event))
-                    .push_bind(move_event_name)
-                    .push_bind(event.event.move_event_contents())
-                    .push_bind(sender)
-                    .push_bind(
-                        event
-                            .event
-                            .recipient_serialized()
-                            .expect("Cannot serialize"),
-                    );
-            });
-
-            let res = query_builder
-                .build()
-                .execute(&self.pool)
-                .await
-                .map_err(convert_sqlx_err)?;
-
-            rows_affected += res.rows_affected();
+        let mut events_by_tx: BTreeMap<TransactionDigest, Vec<EventEnvelope>> = BTreeMap::new();
```

Comment:
I think there is a much better way to create event numbers than here within add_events.  What I don't like about using a BTreeMap and splitting batches into one per transaction:
1. It destroys the advantage of batching event writes from multiple TXes (the average number of events per transaction is barely above 1 right now, so basically it nullifies the advantage of batch writes)
2. It adds significant overhead.  A BTreeMap is an expensive allocation and so is inserting into a map3. 
3. Every implementation of the event store would have to do something like this. 4.4.  

It seems this was done just so that we could assign event_num starting with a zero for each TX.  
Instead, I think the event_num should be added before the event_store itself.  For example, you can have code that numerates the event_num in post_process_one_tx().

That would simplify the code in add_events() significantly and make it much faster.

Reply:
Assigning event_num at event creation time is indeed much simpler.